File: ts/src/parkingManager/parkingManager.ts
/** @format */

import {ParkingBoy} from '../parkingBoy/parkingBoy'

export class ParkingManager {
    public constructor(private parkingBoys: ParkingBoy[]) {}

    private findOneValidParkingBoy() {
        return (
            this.parkingBoys
                .filter(parkingBoy => parkingBoy.checkAvailableLot())
                .sort(() => Math.floor(Math.random() * this.parkingBoys.length))[0] || null
        )
    }

    public findOneValidParkingLot() {
        const parkingBoy = this.findOneValidParkingBoy()
        return parkingBoy && (parkingBoy.findOneValidParkingLot() || null)
    }
}

File: kt/src/main/kotlin/com/github/scglwsj/parking/manager/ParkingManager.kt
package com.github.scglwsj.parking.manager

import com.github.scglwsj.parking.boy.ParkingBoy
import com.github.scglwsj.parking.lot.ParkingLot

class ParkingManager(private val parkingBoys: List<ParkingBoy>) {
    private fun findOneValidParkingBoy(): ParkingBoy? =
        parkingBoys.filter { it.checkAvailableLot() }.minBy { (0..parkingBoys.size).random() }

    fun findOneValidParkingLot(): ParkingLot? = findOneValidParkingBoy()?.findOneValidParkingLot()
}
